cmd/binlog-collector/collector/collector.go

```diff
@@ -0,0 +1,201 @@
+	if err != nil {
+		return errors.Wrap(err, "new db connection")
+	}
+	defer c.closeDB()
+
+	err = c.CollectBinLogs()
+	if err != nil {
+		return errors.Wrap(err, "collect binlog files:")
+	}
+
+	return nil
+}
+
+func (c *Collector) newDB() error {
+	host, err := c.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, c.pxcUser, c.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	c.db = pxc
+
+	return nil
+}
+
+func (c *Collector) closeDB() error {
+	return c.db.Close()
+}
+
+func (c *Collector) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+c.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost + "." + c.pxcServiceName, nil
+}
+
+func (c *Collector) CollectBinLogs() error {
+	// get last uploaded binlog file name
+	binlogName, err := c.db.GetBinLogName(c.lastSet)
+	if err != nil {
+		return errors.Wrap(err, "get binlog name by set")
+	}
+	list, err := c.db.GetBinLogList()
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+
+	upload := false
+	// if there are no uploaded files we going to upload every binlog file
+	if len(binlogName) == 0 {
+		upload = true
+	}
+
+	for _, binlog := range list {
+		if binlog == binlogName { // this check is for uploading starting from needed file
+			upload = true
+		}
+		if upload {
+			err = c.manageBinlog(binlog)
+			if err != nil {
+				return errors.Wrap(err, "manage binlog")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Collector) manageBinlog(binlog string) error {
+	set, err := c.db.GetGTIDSet(binlog)
+	if err != nil {
+		return errors.Wrap(err, "get GTID set")
+	}
+	var setBuffer bytes.Buffer
+	setBuffer.WriteString(set)
+
+	if setBuffer.Len() == 0 {
+		return nil
+	}
+
+	cmd := exec.Command("mysqlbinlog", "-R", "-h"+c.db.GetHost(), "-u"+c.pxcUser, "-p"+os.ExpandEnv("$PXC_PASS"), binlog)
+	out, err := cmd.StdoutPipe()
+	if err != nil {
+		return errors.Wrap(err, "get stdout pipe")
+	}
+
+	errOut, err := cmd.StderrPipe()
+	if err != nil {
+		return errors.Wrap(err, "get stderr pipe")
+	}
+	err = cmd.Start()
+	if err != nil {
+		return errors.Wrap(err, "run mysqlbinlog command")
+	}
+	defer cmd.Wait()
+
+	err = c.storage.PutObject(binlog, out)
+	if err != nil {
+		return errors.Wrap(err, "put binlog object")
+	}
+
+	var stdErr bytes.Buffer
+	stdErr.ReadFrom(errOut)
+	if stdErr.Bytes() != nil && strings.TrimRight(stdErr.String(), "\n") != db.UsingPassErrorMessage {
```

Comment:
use `ioutil.ReadAll(errOut)` instead of buffer
btw, there is [bytes.TrimRight()](https://pkg.go.dev/bytes#TrimRightl)

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/622/commits/c2c0f622d3137e742c2823df2fad5c0260e04428